chore: refactor to implement home-grown valhalla's filesystem methods with std equivalent

nilsnolde opened the pull request:
This is preparation to remove `valhalla/filesystem.h` in favor of C++17 `std::filesystem` in a next step. Luckily that's not a big deal, as Valhalla's API is almost equivalent to std's API (guess both are derived from boost?). Not sure yet what else this will need, in a separate branch I'm ripping out our filesystem, so this will get updated over time likely:

- put `valhalla::filesystem::last_write_time` to a new `valhalla/filesystem_utils.h` which can be used to extend missing API. Actually `last_write_time` is defined in C++17, but what's not well defined is its `__file_clock` which is only a part of C++20. The problem is that this `__file_clock`'s epoch is implementation defined, not standardized, and e.g. Linux has a epoch of smth like year 2147 or so.

nilsnolde commented on src/baldr/incident_singleton.h:
```diff
@@ -195,23 +196,23 @@ struct incident_singleton_t {
-    filesystem::path inc_dir;
+    std::filesystem::path inc_log_path(config.get<std::string>("incident_log", ""));
+    std::filesystem::path inc_dir;
+    bool inc_dir_exists = false;
```
just a little refactor, bcs `std::filesystem::path` can't be initialized like it's used here.

nilsnolde commented on valhalla/filesystem_utils.h:
```diff
@@ -0,0 +1,22 @@
+namespace std::filesystem {
+inline std::time_t last_write_time_t(const std::filesystem::path& p) {
+  // note, in C++20 there's a proper chrono::clock_cast so we can use filesystem::last_write_time
+  struct stat s;
+  if (stat(p.c_str(), &s) != 0)
+    throw std::runtime_error("could not stat " + p.string());
+  return FS_MTIME(s);
```
this is basically the same as the `last_write_time` function in the valhalla::filesystem, bcs there's some problems with pre-c++20 `std::filesystem` and its clock

nilsnolde commented on valhalla/filesystem_utils.h:
```diff
@@ -0,0 +1,22 @@
+#define FS_MTIME(st_stat) st_stat.st_mtim.tv_sec
+#endif
+
+namespace std::filesystem {
```
probably not a good idea? I did it more out of a lack of other good namespace ideas, but maybe the main `valhalla` namespace is good enough, rather than extending a std class?

nilsnolde replied:
though I doubt `std::filesystem` will ever implement `last_write_time_t` or `save` (it's designed to _not_ modify or create files AFAIK; well, other than deleting them)

nilsnolde commented on valhalla/filesystem_utils.h:
```diff
@@ -0,0 +1,83 @@
+typename std::enable_if<has_data<Container>::value, bool>::type inline save(
+    const std::string& fpath,
+    const Container& data = {}) {
+  if (fpath.empty())
```
this was quite messy code. I didn't clean up too much, could do a bit more. this was fairly painful to refactor

nilsnolde replied:
did a bit more cleanup and changed the arg to `std::filesystem::path` to avoid a bunch of string allocations below

nilsnolde commented on src/skadi/sample.cc:
```diff
@@ -604,25 +605,30 @@ void sample::cache_initialisation(const std::string& data_source) {
 
   // messy but needed
```
it is messy and it's not needed, but works still, so didn't touch it

nilsnolde commented on src/skadi/sample.cc:
```diff
@@ -604,25 +605,30 @@ void sample::cache_initialisation(const std::string& data_source) {
     if (data && data->second != format_t::UNKNOWN) {
-      if (!cache_->insert(data->first, f, data->second)) {
-        LOG_WARN("Corrupt elevation data: " + f);
+      if (!cache_->insert(data->first, fp_str, data->second)) {
```
on windows implicit conversion of filesystem::path to str doesn't work apparently, needed to do it explicitly

kinkard commented on valhalla/tyr/actor.h:
```diff
@@ -1,14 +1,17 @@
-#include <valhalla/proto/api.pb.h>
-
-#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/ptree_fwd.hpp>
```
I hardly imagine how `actor.h` can be used without including `#include <boost/property_tree/ptree.hpp>` because of ctor argument `actor_t` takes, so this change only improves includes for files that don't use `actor_t` at all and where removing inclusion of the `actor.h` would be more beneficial.

nilsnolde replied:
yeah true, I'll revert these changes

nilsnolde replied:
reverted these

nilsnolde commented on src/baldr/graphtile.cc:
```diff
@@ -179,29 +183,29 @@ GraphTile::GraphTile() = default;
     if (file.fail())
       success = false;
-    int err = std::rename(tmp_location.c_str(), disk_location.c_str());
+    int err = std::rename(tmp_location.string().c_str(), disk_location.c_str());
```
>  should live long enough here, but it looks like absolutely redundant thing.

would've loved to, but windows somehow didn't like it :disappointed:

kinkard commented on valhalla/filesystem_utils.h:
```diff
@@ -15,6 +15,32 @@
+  // is finished deleting. In this exact moment, if thread 2 calls
+  // stat() on the file object in question it will get 0 which means
+  // "file object exists". We kinda want thread 2 to wait for the
+  // file object to truly go out of existence. Hence, this spin loop.
```
I would say that this the problem at job orchestration level - don't remove the same file from multiple threads, meaning that the implementation should not try to solve this problem as it could not be solved at this level.
Especially taking in account that `remove` actually unlinks the name from the content and if that file is still open by some process its content would not be deleted (so if some thread acquired fd of the file and then go sleep because of being preempted...)... and all other funny stuff that comes from the nature of the filesystem.

nilsnolde replied:
I mentioned further above, we can safely remove the concurrency requirements for filesystem stuff I think. IMO it'd still be fine if we really really needed it and we can make it work for ourselves. but all the better if we don't.

nilsnolde replied:
this is also removed

kinkard commented on valhalla/filesystem_utils.h:
```diff
@@ -0,0 +1,116 @@
+  @param p The directory to recursively remove
+  @returns The number of files & directories which were removed (if any)
+*/
+inline std::uintmax_t remove_all(const std::filesystem::path& p) {
```
Btw. I saw you've mentioned that this function is only called from tests. Maybe it is simpler just to remove this functionality at all?

nilsnolde replied:
> Btw. I saw you've mentioned that this function is only called from tests.

That's the only place I changed the usage so far, just to test it. The other usages are still based on the original one. That being said: I just had a look where it's used and none of the uses are in any multithreaded code, same for `std::filesystem::create_directories` and both of which are tested for working concurrently. I guess the best way is to simply use the normal std methods and remove that test. Damn, should've looked at it before doing these gymnastics.. Thanks for the hint!

nilsnolde commented:
did another round of edits @kinkard , thanks for the review. I removed the concurrency requirements (weren't needed in the current data tools) and cleaned up the existing stuff a bit more.

nilsnolde commented on src/baldr/graphtile.cc:
```diff
@@ -176,32 +180,33 @@ GraphTile::GraphTile(const std::string& tile_dir,
 GraphTile::GraphTile() = default;
 
-void GraphTile::SaveTileToFile(const std::vector<char>& tile_data, const std::string& disk_location) {
+void GraphTile::SaveTileToFile(const std::vector<char>& tile_data,
```
this method isn't used anymore it seems. it was introduced in https://github.com/valhalla/valhalla/pull/2026 for `GraphTile::CacheTileURL` but that code was ripped out at some point later. might as well remove it?

thomthom commented:
Have you tested this with files with unicode paths? I'm not familiar with the valhalla filesystem API you are replacing. But I've had trouble with `std::filesystem` on Windows when I came to unicode paths. Even C++20's `u8path` wasn't something that worked well cross platform.

nilsnolde commented:
reviewed as part of #5321 , closing here
